src/mqtt/buffer_ref.h

```diff
@@ -0,0 +1,278 @@
+	 */
+	buffer_ref& operator=(const char* cstr) {
+		static_assert(sizeof(char) == sizeof(T), "can only use C arr with char or byte buffers");
+		data_.reset(new blob(reinterpret_cast<const value_type*>(cstr), strlen(cstr)));
+		return *this;
+	}
+	/**
+	 * Copy another type of buffer reference to this one.
+	 * This can copy a buffer of different types, provided that the size of
+	 * the data elements are the same. This is typically used to convert
+	 * from char to byte, where the data is the same, but the interpretation
+	 * is different. Note that this copies the underlying buffer.
+	 * @param rhs A reference to a different type of buffer.
+	 * @return A reference to this object.
+	 */
+	template <typename OT>
+	buffer_ref& operator=(const buffer_ref<OT>& rhs) {
+		static_assert(sizeof(OT) == sizeof(T), "Can only assign buffers if values the same size");
+		data_.reset(new blob(reinterpret_cast<const value_type*>(rhs.data()), rhs.size()));
+		return *this;
+	}
+	/**
+	 * Clears the reference to nil.
+	 */
+	void reset() { data_.reset(); }
+	/**
+	 * Get a const pointer to the data buffer.
+	 * @return A pointer to the data buffer.
+	 */
+	const value_type* data() const { return data_ ? data_->data() : nullptr; }
+	/**
+	 * Get the size of the data buffer.
+	 * @return The size of the data buffer.
+	 */
+	size_t size() const { return data_ ? data_->size() : 0; }
+	/**
+	 * Get the size of the data buffer.
+	 * @return The size of the data buffer.
+	 */
+	size_t length() const { return data_ ? data_->length() : 0; }
+	/**
+	 * Get the data buffer as a string.
+	 * @return The data buffer as a string.
+	 */
+	//blob to_string() const { return data_ ? (*data_) : blob(); }
+	const blob& to_string() const { return *data_; }
+	/**
+	 * Get the data buffer as NUL-terminated C string.
+	 * Note that the reference must be set to call this function.
+	 * @return The data buffer as a string.
+	 */
+	const char* c_str() const { return data_->c_str(); }
+	/**
+	 * Gets a shared pointer to the (const) data buffer.
+	 * @return A shared pointer to the (const) data buffer.
+	 */
+	const pointer_type& ptr() const { return data_; }
+	/**
+	 * Gets elemental access to the data buffer (read only)
+	 * @param i The index into the buffer.
+	 * @return The value at the specified index.
+	 */
+	const value_type& operator[](size_t i) const { return (*data_)[i]; }
+};
+
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const buffer_ref<T>& buf) {
+	if (!buf.empty())
+		os.write(buf.data(), buf.size());
+	return os;
+}
+
+/////////////////////////////////////////////////////////////////////////////
+
+/**
+ * A refernce to a text buffer.
+ */
+using string_ref = buffer_ref<char>;
+
+/**
+ * A reference to a binary buffer.
+ * Note that we're using char for the underlying data type to allow
+ * efficient moves to and from std::string's. Using a separate type
+ * indicates that the data may be arbitrary binary.
+ */
+using binary_ref = buffer_ref<char>;
```

Comment:
Using a normal string reference (const std::string&) would not provide ownership and lifetime management. That's what the shared pointer gives us. The key is to make it a shared pointer to a _const_ object so that we can create as many copies as we want and pass them around different threads without worrying that anyone would change the object; no locks required.

We could just use a shared pointer directly in the classes, like:
```
class message {
    //...
    std::shared_ptr<const std::string> payload_;
    //...
```
Wrapping this in a class gives us a few advantages:

- It seemed to offer a quicker drop-in replacement from what was already in the code, so fewer changes were required through the rest of the code (i.e. laziness).
- It provides a means to change the implementation later without changing the public API (which will be locked for the 1.x versions here on out).
- With a number of implicit constructors, there is great flexibility in the API, and how we convert from different types.

For example, if the `connect_options` class has the member:
```
void set_user_name(string_ref userName);
```
It could be called with:
```
opt.set_user_name("myuser");
opt.set_user_name(std::string("myuser"));
opt.set_user_name(new std::string("myuser"));
```
and so on. 

As for the `binary_ref`, you have perhaps seen me flip flop back and forth on this one. Using unsigned bytes seems the most "correct" to hold binary data, but I've come up with a few reasons to stick with char strings:
- There is no language support for binary literals, but you can make them as char strings, like, `"\x1\x2\x3"`. It's a little ugly, but possible. This, however gives a const _char_ pointer. Barring that we would need to consider aC++11 user-defined literal which might do nothing more than cast the result.
- std::string can hold any binary data. We just don't want to use the type directly (without at least a typecast) because it's confusing about what it _might_ be holding. But it works for binary or char data.
- Although we want to support binary, a lot of users will likely use text payloads (JSON, etc). So we want something that is convenient for C and C++ strings.
- If we use a basic_string<byte> we lose move semantics from a string (basic_string<char>) since they are different types. This is often a nice optimization.

But I'm not completely settled on this. We should experiment with it. That's another reason why I went with the wrapper template for `string_ref` and `binary_ref`. We can control conversions internally without subjecting users to ugly casts.